source/extensions/filters/http/oauth2/oauth_client.h

```diff
@@ -91,6 +91,7 @@ class OAuth2ClientImpl : public OAuth2Client, Logger::Loggable<Logger::Id::upstr
     auto request = Http::Utility::prepareHeaders(uri_);
     request->headers().setReferenceMethod(Http::Headers::get().MethodValues.Post);
     request->headers().setContentType(Http::Headers::get().ContentTypeValues.FormUrlEncoded);
+    request->headers().addCopy("accept", "application/json");
```

Comment:
Seems like:

```
    request->headers().setReferenceMethod(Http::Headers::get().MethodValues.Post);
    request->headers().setReferenceContentType(
        Http::Headers::get().ContentTypeValues.FormUrlEncoded);
    request->headers().addReference(Http::CustomHeaders::get().Accept,
                                    Http::Headers::get().ContentTypeValues.Json);

```

While you're here probably we can set that `setContentType` to `setReferenceContentType`.

And we can add test to: test/extensions/filters/http/oauth2/oauth_test.cc

e.g. in `TEST_F(OAuth2ClientTest, RequestAccessTokenSuccess) {`.

```cpp
  EXPECT_CALL(cm_.thread_local_cluster_.async_client_, send_(_, _, _))
      .WillRepeatedly(Invoke([&](Http::RequestMessagePtr& message, Http::AsyncClient::Callbacks& cb,
                                 const Http::AsyncClient::RequestOptions&)
                                 -> Http::AsyncClient::Request* {
        EXPECT_FALSE(message->headers().get(Http::CustomHeaders::get().Accept).empty());
        EXPECT_EQ(
            message->headers().get(Http::CustomHeaders::get().Accept)[0]->value().getStringView(),
            Http::Headers::get().ContentTypeValues.Json);
        callbacks_.push_back(&cb);
        return &request_;
      }));
```

Reply:
Thanks for the help!

I applied that almost exactly, but I had to change the Accept header value count assertion (`empty()`) from `ASSERT_FALSE` to `EXPECT_FALSE`, otherwise it wouldn't compile, probably because it's a lambda. But that also means the testing code is going to do an out-of-bounds array access in the line below if the accept header is not set. 


https://github.com/envoyproxy/envoy/blob/4d9fa702209a00cd1d9b407048af075d5c620435/test/extensions/filters/http/oauth2/oauth_test.cc#L85-L87

Do you have a better idea how to do this?

Sorry, for not getting to this sooner. I took me a while to find a machine that is big enough to run the Envoy unit tests without crashing :P
![image](https://user-images.githubusercontent.com/21371810/103351613-0d1c9100-4aa4-11eb-89d8-f0884cd8ceb1.png)

Reply:
Would this work using short-circuit evaluation?

```c++
EXPECT_FALSE(
    message->headers().get(Http::CustomHeaders::get().Accept).empty() ||
    message->headers().get(Http::CustomHeaders::get().Accept)[0]->value().getStringView() !=
        Http::Headers::get().ContentTypeValues.Json);
```